File: bongf/python/level2/string_compression.py
## 앞에 숫자가 무조건 한자리수라고 생각했는데, a * 10인 경우, 앞의 두글자를 확인해줘야 한다.! 
def solution(s):

    ans_list = []
    for i in range(1, len(s)+1) :
        t = s
        temp = []
        while(len(t) >= i) :
            x = t[0:i]
            t = t[i:len(t)]
            if(len(temp) != 0 and (temp[-1] == x)):
                temp[-2] = str(int(temp[-2]) + 1) 
            else : 
                temp.append('1')
                temp.append(x)

        if(t != '') :
            temp.append(t)

        while('1' in temp):
            temp.remove('1')

        ans_list.append("".join(temp))

    ans_list.sort(key= lambda x : len(x))
    return len(ans_list[0])

def compress(text, tok_len):
    words = [text[i:i+tok_len] for i in range(0, len(text), tok_len)]
    print(words)
    print(words[1:] + [''])
    res = []
    cur_word = words[0]
    cur_cnt = 1
    # 아래와 같이 for문을 돌리면 바로 앞과 뒤의 인자들을 비교할 수 있다. 
    for a, b in zip(words, words[1:] + ['']):
        print(a, b)
        if a == b:
            cur_cnt += 1
        else:
            res.append([cur_word, cur_cnt])
            cur_word = b
            cur_cnt = 1

    print(res)
    # return 해당 문자열의 길이로 해줘서 최종 값에서 min 으로 출력하게 해줌         
    return sum(len(word) + (len(str(cnt)) if cnt > 1 else 0) for word, cnt in res)

def solution2(text):
    # 절반을 넘어가면 어쩄든 맨 앞의 문자 + 뒤의 문자니까 전체 문자열 한 것과 똑같다!  
    # 그래서 절반까지의 list를 만들고 거기에 전체 길이를 하나 append 해서 길이에 대한 list를 만든 것 
    # list(range(1, int(len(text)/2) + 1)) + [len(text)]

    return min(compress(text, tok_len) for tok_len in list(range(1, int(len(text)/2) + 1)) + [len(text)])

# print(solution("aaaaaaaaaaaaaaa"))
print(solution2("aabbaccc"))
# print(solution("ababcdcdababcdcd"))
# print(solution("abcabcdede"))
# print(solution("abcabcabcabcdededededede"))
# print(solution("xababcdcdababcdcd"))

File: bongf/java/src/level2/StringCompression.java
package level2;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class StringCompression {
    public static void main(String[] args) {
        StringCompression sc = new StringCompression();
        System.out.println(sc.solution("aabbaccc"));
        System.out.println(sc.solution("ababcdcdababcdcd"));
        System.out.println(sc.solution("abcabcdede"));
        System.out.println(sc.solution("abcabcabcabcdededededede"));
        System.out.println(sc.solution("xababcdcdababcdcd"));
    }

    public int solution(String s) {
        List<Integer> strings = new ArrayList<>();
        for (int i = 1; i < s.length() / 2 + 1; i++) {
            strings.add(compression(i, s));
        }
        strings.add(compression(s.length(), s));
        Collections.sort(strings);
        return strings.get(0);
    }

    public int compression(int length, String s) {
        List<String> stringList = new ArrayList<>();
        for (int i = 0; i < s.length(); i += length) {
            if (i + length > s.length()) {
                stringList.add(s.substring(i));
            } else {
                stringList.add(s.substring(i, i + length));
            }
        }

        List<String> copyList = new ArrayList(stringList);
        copyList.remove(0);
        // 이거 안해주고 마지막 전의 것과 마지막 것만 비교하는 것에서 끝냈더니 add 안해주고 count 만 증가시키고 끝남
        copyList.add("");
        int curCount = 1;
        List<String> returnStringList = new ArrayList<>();
        for (int i = 0; i < stringList.size(); i++) {
            if (stringList.get(i).equals(copyList.get(i))) {
                curCount++;
            } else {
                if (curCount != 1) {
                    returnStringList.add(String.valueOf(curCount));
                }
                returnStringList.add(stringList.get(i));
                curCount = 1; 
            }
        }
        return String.join("", returnStringList).length();
    }

    public int solution2(String s) {
        int answer = 0;

        for(int i=1; i<=(s.length()/2)+1; i++){
            int result = getSplitedLength(s, i, 1).length();
            answer = i==1 ? result : (answer>result?result:answer);
        }

        return answer;
    }

    public String getSplitedLength(String s, int n, int repeat){
        if(s.length() < n) return s;
        String result = "";
        String preString = s.substring(0, n);
        String postString = s.substring(n, s.length());

        // 불일치 -> 현재까지 [반복횟수 + 반복문자] 조합
        if(!postString.startsWith(preString)){
            if(repeat ==1) return result += preString + getSplitedLength(postString, n, 1);
            return result += Integer.toString(repeat) + preString + getSplitedLength(postString, n, 1);
        }

        return result += getSplitedLength(postString, n, repeat+1);
    }
}
